### PR TITLE
[functorch] add functorch functional_call, update tests to test this

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88850
* __->__ #89213

### CODE REVIEW

**Comment by samdow** [comment]
@zou3519 stack now starts here. It's ready for review!

File: torch/_functorch/functional_call.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
```
Comment (zou3519) [inline torch/_functorch/functional_call.py:5]: nit: don't think we need the header anymore. It's not used for pytorch/pytorch but is required for all the other repos (pytorch/functorch, etc)

File: test/functorch/test_eager_transforms.py [hunk 2]
```diff
@@ -146,6 +147,29 @@ def test_argnums_reorders(self):
         res = _slice_argnums(args, (1, 0))
         self.assertEqual(res, (args[1], args[0]))
 
+def _get_weights_and_functional_call(net, mechanism):
+    if mechanism == "make_functional":
+        return make_functional(net)
+    else:
+        # this makes it so the function from make_functional and this call have the same signature
```
Comment (zou3519) [inline test/functorch/test_eager_transforms.py:154]: nit: assert mechanism == "functional_call"

File: torch/_functorch/functional_call.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import torch.nn as nn
+from torch import Tensor
+
+
+def functional_call(
+    module: 'torch.nn.Module',
+    parameter_and_buffer_dicts: Union[Dict[str, Tensor], Tuple[Dict[str, Tensor]]],
+    args: Union[Any, Tuple],
+    kwargs: Dict[str, Any] = None,
+):
```
Comment (zou3519) [inline torch/_functorch/functional_call.py:18]: - I assume the docs are coming later?
- did you want nn.utils.stateless.functional_call to also have this API, or does nn.utils.stateless.functional_call have a different API (and we deprecate it?)

File: torch/_functorch/functional_call.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        for key in keys:
+            if keys.count(key) > 1:
+                raise ValueError(f"{key} appeared in multiple dictionaries; behavior of functional call is ambiguous")
+
+        parameters_and_buffers = {k: v for d in parameter_and_buffer_dicts for k, v in d.items()}
+
+    return nn.utils.stateless.functional_call(module, parameters_and_buffers, args, kwargs)
+
+
+def params_and_buffers_no_grad_tracking(module: 'torch.nn.Module'):
```
Comment (zou3519) [inline torch/_functorch/functional_call.py:31]: To check: `params_and_buffers_no_grad_tracking` is the replacement for make_functional's disable_autograd_tracking flag?

`params_and_buffers_disable_autograd_tracking` is more consistent with the original name for make_functional. Though, I don't really like having this API because it's a one-liner. It made sense for make_functional because make_functional is supposed to return a functional module and some parameters that can be used with said module and the user tells make_functional if they're planning on using regular autograd or functorch grad with it. Thoughts?

File: test/functorch/test_eager_transforms.py [hunk 5]
```diff
@@ -146,6 +147,29 @@ def test_argnums_reorders(self):
...
+        weights = {k: v for k, v in net.named_parameters()}
+        return net_func, weights
+
+def _get_weights_and_functional_call_with_buffers(net, mechanism):
+    if mechanism == "make_functional":
+        return make_functional_with_buffers(net)
+    else:
+        # this makes it so the function from make_functional and this call have the same signature
+        def net_func(weights, buffers, data):
+            return functional_call(net, {**weights, **buffers}, (data,))
```
Comment (zou3519) [inline test/functorch/test_eager_transforms.py:167]: should use functional_call(net, (weights, buffers), (data,)) to test that path of function_call

File: torch/_functorch/functional_call.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+
+def functional_call(
+    module: 'torch.nn.Module',
+    parameter_and_buffer_dicts: Union[Dict[str, Tensor], Tuple[Dict[str, Tensor]]],
+    args: Union[Any, Tuple],
+    kwargs: Dict[str, Any] = None,
+):
+    parameters_and_buffers = parameter_and_buffer_dicts if isinstance(parameter_and_buffer_dicts, dict) else {}
+    if isinstance(parameter_and_buffer_dicts, tuple):
```
Comment (zou3519) [inline torch/_functorch/functional_call.py:20]: Should we assert that parameter_and_buffer_dicts is a tuple of two things here? Or is it fine for it to have arbitrary number of dicts? (at which point we should rename it `member_dicts` or something)

File: torch/_functorch/functional_call.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+from typing import Dict, Union, Any, Tuple
+
+import torch
+import torch.nn as nn
+from torch import Tensor
+
+
+def functional_call(
+    module: 'torch.nn.Module',
+    parameter_and_buffer_dicts: Union[Dict[str, Tensor], Tuple[Dict[str, Tensor]]],
```
Comment (zou3519) [inline torch/_functorch/functional_call.py:15]: Tuple[Dict[str, Tensor]], Dict[str, Tensor]]] to demonstrate it's a tuple of two things if that's what was intended

**Review by zou3519 — COMMENTED** [review]
Code looks reasonable to me. I had some high-level questions (that I might have asked on the other PRs, sorry if they're redundant):
- should the new functional_call API be the same as the nn.stateless.utils.functional_call API, or is the plan that we'll deprecate the nn.stateless.utils.functional_call immediately?
- are we exposing the new functional_call in `functorch`, or should it just be exposed in torch.func?
- should we have params_and_buffers_disable_autograd_tracking ? Or should we just warn somewhere in the functional_call docs and show an example with .detach() ?

File: functorch/__init__.py [hunk 8]
```diff
@@ -28,4 +28,6 @@
     FunctionalModuleWithBuffers,
 )
 
+from torch._functorch.functional_call import functional_call, params_and_buffers_no_grad_tracking
+
```
[the remark this answers is not in the document]
Comment (samdow) [reply functorch/__init__.py:32]: Yes it was just an intermediate until torch.func was available. Also as a reminder to us to expose it when we move it to torch.func

File: torch/_functorch/functional_call.py (hunk 3 above)
Comment (samdow) [reply torch/_functorch/functional_call.py:18]: Ahh thanks for the catch I forgot to move the docs over

Re: `nn.utils.stateless.functional_call`, I think we should deprecate it and ask users to use torch.func instead since they are the same API. I'll add a PR to do that. Also since it was technically BC-breaking to change this since previously parameters_and_buffers must have been a single dictionary

File: torch/_functorch/functional_call.py (hunk 4 above)
Comment (samdow) [reply torch/_functorch/functional_call.py:31]: Yep this was a replacement for that. I'll remove it and add a note to functional_call instead

File: torch/_functorch/functional_call.py (hunk 6 above)
Comment (samdow) [reply torch/_functorch/functional_call.py:20]: Yeah this should be able to take in an arbitrary number of dictionaries (I haven't written it out, but I thought it might help integration with tensordict). I'll change the signature and name to reflect this

File: functorch/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply functorch/__init__.py:32]: sgtm

**Comment by zou3519** [comment]
Btw, torch.func exists now, so let's put functional_call there

**Comment by samdow** [comment]
> Btw, torch.func exists now, so let's put functional_call there

Done! Let me know if there's anything else!

File: torch/_functorch/functional_call.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+    Returns:
+        Any: the result of calling ``module``.
+    """
+    parameters_and_buffers = parameter_and_buffer_dicts if isinstance(parameter_and_buffer_dicts, dict) else {}
+    if isinstance(parameter_and_buffer_dicts, tuple):
+        keys = [parameter_and_buffer.keys() for parameter_and_buffer in parameter_and_buffer_dicts]
+        for key in keys:
+            if keys.count(key) > 1:
+                raise ValueError(f"{key} appeared in multiple dictionaries; behavior of functional call is ambiguous")
```
Comment (zou3519) [inline torch/_functorch/functional_call.py:75]: Sorry I am a bit confused about what is going on here.
- keys is a List[List[str]], but don't we want a List[str]?
- Also, doing keys.count(key) seems like O(n) work, so the entire check seems O(n^2), which is bad if there's a lot of keys, right? We can short circuit this if there is a single dictionary.

We should add a test where we have dictionaries with duplicate keys and assert that this error gets raised

File: test/functorch/test_eager_transforms.py [hunk 10]
```diff
@@ -146,6 +149,30 @@ def test_argnums_reorders(self):
         res = _slice_argnums(args, (1, 0))
         self.assertEqual(res, (args[1], args[0]))
 
+def _get_weights_and_functional_call(net, mechanism):
```
Comment (zou3519) [inline test/functorch/test_eager_transforms.py:152]: All of the tests for functional_call seem integration-y (they run on a model).

Could we add some basic API unittests to test things like:
- functional_call with a single dict
- functional_call with multiple dicts
- functional_call with args, kwargs
- functional_call's special case where it accepts a single input

This way, if someone modifies torch.func.functional_call, the tests will notice it

File: torch/_functorch/functional_call.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+from torch import Tensor
+from torch._functorch.utils import exposed_in
+
+
+@exposed_in("torch.func")
+def functional_call(
+    module: 'torch.nn.Module',
+    parameter_and_buffer_dicts: Union[Dict[str, Tensor], Tuple[Dict[str, Tensor], ...]],
+    args: Union[Any, Tuple],
+    kwargs: Dict[str, Any] = None,
```
Comment (zou3519) [inline torch/_functorch/functional_call.py:14]: Just to check: Did the weight tying update for stateless.functional_call happen already? If so, should functional_call have an option to toggle the weight tying behavior?

**Review by zou3519 — COMMENTED** [review]
I'm a bit confused about the duplicate keys check, but other than that, this LGTM. I had some minor comments on nits and suggestions to improve the testing

File: docs/source/func.api.rst [hunk 12]
```diff
@@ -20,3 +20,43 @@ Function Transforms
...
+
+    functional_call
+
+If you're looking for information on fixing Batch Norm modules, please follow the
+guidance here
+
+.. toctree::
+   :maxdepth: 1
+
+   batch_norm
```
[the remark this answers is not in the document]
Comment (samdow) [reply docs/source/func.api.rst]: Good catch! Moved it over but it still refers to `functorch.experimental.replace_all_batch_norm_modules_` I was going to add a PR at the top to fix this but happy to squash it in here if that's preferred

File: torch/_functorch/functional_call.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+        Example::
+
+            >>> mod = nn.Linear(1, 1)
+            >>> d = {k: v.detach() for k, v in mod.named_parameters()}
+            >>> grad(lambda x: functional_call(mod, d, x), torch.randn((1, 1)))  # doesn't tracks grads for params
+
+    Args:
+        module (torch.nn.Module): the module to call
+        parameters_and_buffers (dict of str and Tensor or tuple of str and Tensors): the parameters that will be used in
```
[the remark this answers is not in the document]
Comment (samdow) [reply torch/_functorch/functional_call.py]: Fixed thanks!

File: torch/_functorch/functional_call.py (hunk 9 above)
Comment (samdow) [reply torch/_functorch/functional_call.py]: Good catch! Added a test for this and made the logic a little easier--basically it now makes a list of all the keys, flattened, and then make that a set. It compares the length of the two and then does the count check to look for the repeated key if they're different lengths

Although this is better at O(n) time to make the list and then set from the list and O(n) space (it will still be O(n^2) in the error case to find the repeated key). As for the short circuit, this will short circuit because we only do the O(n) check when we're passed a tuple. If we get a single dictionary, it will skip this and run as before

File: test/functorch/test_eager_transforms.py (hunk 10 above)
Comment (samdow) [reply test/functorch/test_eager_transforms.py:152]: Added tests to `test_stateless.py`

File: torch/_functorch/functional_call.py (hunk 11 above)
Comment (samdow) [reply torch/_functorch/functional_call.py:14]: #90477 hasn't been approved yet. I can add it in there if this lands first

File: torch/_functorch/functional_call.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+            >>> x = torch.randn((1, 1))
+            >>> print(x)
+            >>> functional_call(mod, a, x)  # same as x
+            >>> print(mod.weight)  # same as before functional_call
+
+        Example::
+
+            >>> mod = nn.Linear(1, 1)
+            >>> d = {k: v.detach() for k, v in mod.named_parameters()}
+            >>> grad(lambda x: functional_call(mod, d, x), torch.randn((1, 1)))  # doesn't tracks grads for params
```
[the remark this answers is not in the document]
Comment (samdow) [reply torch/_functorch/functional_call.py]: Added end-to-end example and moved this down to be a part of the note on not tracking. Here's what the ending looks like, rendered

Screen Shot 2022-12-27 at 3 25 42 PM

File: docs/source/func.api.rst [hunk 15]
```diff
@@ -20,3 +20,43 @@ Function Transforms
...
+    assert jacobian.shape == (3, 3)
+
+However, if you want to do something like compute a jacobian over the parameters
+of the model, then there needs to be a way to construct a function where the
+parameters are the inputs to the function.
+That's what :func:`functional_call` is for:
+it accepts an nn.Module, the transformed ``parameters``, and the inputs to the
+Module's forward pass. It returns the value of running the Module's forward pass
+with the replaced parameters.
+
```
Comment (zou3519) [inline docs/source/func.api.rst:49]: nit: I know this was how it was like in the original docs, but it would be nice to see the jacrev over parameters example here so that the user doesn't need to click into the functional_call docs.

File: docs/source/func.batch_norm.rst [hunk 16]
```diff
@@ -0,0 +1,58 @@
+Patching Batch Norm
+===================
+
+What's happening?
+-----------------
+Batch Norm requires in-place updates to running_mean and running_var of the same size as the input.
+Functorch does not support inplace update to a regular tensor that takes in a batched tensor (i.e.
```
Comment (zou3519) [inline docs/source/func.batch_norm.rst:7]: nit: vmap? And link section from the UX limitations doc if possible (there is some sphinx syntax to link it)

File: docs/source/func.batch_norm.rst [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+``regular.add_(batched)`` is not allowed). So when vmaping over a batch of inputs to a single module,
+we end up with this error
+
+How to fix
+----------
+All of these options assume that you don't need running stats. If you're using a module this means
+that it's assumed you won't use batch norm in evaluation mode. If you have a use case that involves
+running batch norm with vmap in evaluation mode, please file an issue
```
Comment (zou3519) [inline docs/source/func.batch_norm.rst:15]: Are we not going to recommend replacing the batch norm with another normalization layer? That should be good for completeness so the user knows their options

File: test/test_stateless.py [hunk 18]
```diff
@@ -44,43 +45,47 @@ def _run_call_with_mock_module(self, module, device='cpu', prefix=''):
...
         self.assertEqual(x, res)
         # check that the weight remain unmodified
         cur_weight = to_check.l1.weight
         cur_buffer = to_check.buffer
         self.assertEqual(cur_weight, prev_weight)
         self.assertEqual(cur_buffer, prev_buffer)
 
-    def test_functional_call(self):
+    @parametrize("functional_call", [torch.func.functional_call, stateless.functional_call])
+    def test_functional_call(self, functional_call):
```
Comment (zou3519) [inline test/test_stateless.py:57]: nit: What exactly does the test name for this look like? If it has any special characters in it (like `test_functional_call_<function functional_call of torch.func>` then we want to parametrize with subtest to give it a sane name.

e.g.
```
 @parametrize_test('weight_layout', [
     subtest(torch.strided, name='weightStrided'),
     subtest(torch.sparse_coo, name='weightCOO'),
     subtest(torch.sparse_csr, name='weightCSR'),
     subtest(torch.sparse_csc, name='weightCSC'),
     # TODO: addmm: computation on CPU is not implemented for Strided + Strided @ SparseBsr
     # subtest(torch.sparse_bsr, name='weightBSR'),
     # subtest(torch.sparse_bsc, name='weightBSC'),
 ])
 ```

File: test/test_stateless.py [hunk 19]
```diff
@@ -219,11 +228,50 @@ def forward(self, x):
...
+        self.assertEqual(res, res_1)
+
+
+    def test_functional_call_multiple_dicts(self):
+        mod = MockModule()
+        x = torch.rand((1, 1))
+        parameters = {k: torch.ones_like(v) for k, v in mod.named_parameters()}
+        buffers = {k: torch.zeros_like(v) for k, v in mod.named_buffers()}
+        res = torch.func.functional_call(mod, (parameters, buffers), x)
+        self.assertEqual(res, x + 1)
```
Comment (zou3519) [inline test/test_stateless.py:263]: nit:
- would try one example with 3 or more dicts, and another one with zero dicts. I can see a situation in the future where someone just assumes the number of dicts goes up to 2

**Review by zou3519 — APPROVED** [review]
LGTM, some minor nits

File: torch/_functorch/functional_call.py (hunk 11 above)
Comment (zou3519) [reply torch/_functorch/functional_call.py:14]: Sorry, totally missed that PR

File: docs/source/func.api.rst (hunk 15 above)
Comment (samdow) [reply docs/source/func.api.rst:49]: added!

File: docs/source/func.batch_norm.rst (hunk 16 above)
Comment (samdow) [reply docs/source/func.batch_norm.rst:7]: I think the UX limitations hasn't been pulled over yet (probably my bad for not looking at a PR yet. Filed #91509 so I don't forget to do this in a follow up)

File: docs/source/func.batch_norm.rst (hunk 17 above)
Comment (samdow) [reply docs/source/func.batch_norm.rst]: Added language to 1 + 2 on how to replace with GroupNorm. From Opacus' work, generally replacing the norm layer with a different layer is fragile but the version functorch offers (`replace_batch_norm_modules_`) turns off tracking should work since it just updates the layer in place instead of trying to replace the whole layer

File: test/test_stateless.py (hunk 18 above)
Comment (samdow) [reply test/test_stateless.py]: TIL about subtest! Thanks!

File: test/test_stateless.py (hunk 19 above)
Comment (samdow) [reply test/test_stateless.py]: added!
